docs/spec_draft.md

```diff
@@ -941,6 +942,143 @@ produces an implementation-defined value.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.dot_general
+
+### Semantics
+
+Performs general dot product between `lhs` and `rhs` over dimensions specified
+in lhs/rhs batching and contracting dimensions and produces a `result` tensor.
+
+This general form allows dot product between multi-dimensional tensors, and to
+contract multiple dimensions at a time. It also allows to specify the batching
+dimensions, which determines the dimensions to avoid summing during contraciton.
+
+More formally, let:
+* `batching_dims` = [0, 1, ..., size(`lhs_batching_dims`)].
+* LCD = `lhs_contracting_dimensions`.
+* RCD = `rhs_contracting_dimensions`.
+
+$$result[...,\ ip,\ ...] = \sum_{d\ =\ 0}^{size(LCD)} \sum_{(c,\ c')\ =\
+(0,\ 0)}^{(dim(lhs,\ LCD[d]),\ dim(rhs,\ RCD[d]))}
+lhs(...,\ jc,\ ...,\ jq,\ ...)\ *\ rhs(...,\ kc',\ ...,\ kr,\ ...)$$
+
+* If p is in `batching_dims`, j[`lhs_batching_dims`[p]] = ip and
+k[`rhs_batching_dims`[p]] = ip.
+* jq = kr = ip otherwise.
```

Comment:
Proposed rewording:

```
Computes dot products between slices of `lhs` and slices of `rhs` and produces
`result`.

More formally, `result[result_index] = dot_product` where
  * `lhs_result_dimensions = [d for d in axes(lhs) and d not in lhs_batching_dimensions and d not in lhs_contracting_dimensions]`.
  * `rhs_result_dimensions = [d for d in axes(rhs) and d not in rhs_batching_dimensions and d not in rhs_contracting_dimensions]`.
  * `result_batching_index + result_lhs_index + result_rhs_index = result_index`
    where `size(result_batching_index) = size(lhs_batching_dimensions)`,
    `size(result_lhs_index) = size(lhs_result_dimensions)` and
    `size(result_rhs_index) = size(rhs_result_dimensions)`.
  * `transposed_lhs = transpose(lhs, lhs_batching_dimensions + lhs_result_dimensions + lhs_contracting_dimensions)`.
  * `transposed_lhs_slice = slice(result_batching_index + result_lhs_index + [:, ...:])`.
  * `reshaped_lhs_slice = reshape(transposed_lhs_slice, dims(lhs, lhs_contracting_dimensions))`.
  * `transposed_rhs = transpose(rhs, rhs_batching_dimensions + rhs_result_dimensions + rhs_contracting_dimensions)`.
  * `transposed_rhs_slice = slice(result_batching_index + result_rhs_index + [:, ...:])`.
  * `reshaped_rhs_slice = reshape(transposed_rhs_slice, dims(rhs, rhs_contracting_dimensions))`.
  * `dot_product = reduce(
    inputs=[multiply(reshaped_lhs_slice, reshaped_rhs_slice)], 
    init_values=[0],
    dimensions=[0, ..., size(lhs_contracting_dimensions) - 1],
    body=lambda x, y: add(x, y))`.
```

Feedback:
  1. Proposed wording reformulates "general dot product" as "dot products between slices of lhs and slices of rhs", which more clearly summarizes the behavior of the op.
  2. In the last few weeks, we've been gravitating towards expressing ops via other ops if possible, and the proposed formalism goes along the same lines.

Reply:
This can be left for future work, but does this new formulation affect the current interpreter implementation since we want to have a reference implementation for the interpreter?

Reply:
Let's leave resolving this to when we get to the reference implementation. Generally, it is an open question of how the spec is going to be mapped on the interpreter.